### PR TITLE
scheduler: refactor solve in hot-region-scheduler

### PR DESCRIPTION
**Pull request by HunDunDM**
### What problem does this PR solve?

Issue Number: Ref #4949

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

### Release note

```release-note
None
```

### CODE REVIEW

File: server/schedulers/hot_region.go [hunk 1]
```diff
@@ -465,27 +467,31 @@ func (bs *balanceSolver) solve() []*operator.Operator {
 		return nil
 	}
 	bs.cur = &solution{}
+	checkSolution := func() {
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:470]: How about `updateBestSolution`?

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -1012,94 +1018,124 @@ func (bs *balanceSolver) isReadyToBuild() bool {
...
+	switch bs.cur.progressiveRank {
+	case -3:
+		dim = "all"
+	case -2:
+		dim = dimToString(bs.secondPriority)
+	case -1:
+		dim = dimToString(bs.firstPriority)
+	}
+
+	var getOperator func(region *core.RegionInfo, srcStoreID, dstStoreID uint64) (op *operator.Operator, typ string, err error)
```
Comment (lhy1024) [inline server/schedulers/hot_region.go:1040]: How about `createOperator`?

**Review by lhy1024 — APPROVED** [review]

File: server/schedulers/hot_region.go [hunk 3]
```diff
@@ -1012,94 +1018,124 @@ func (bs *balanceSolver) isReadyToBuild() bool {
...
+		schedulerCounter.WithLabelValues(bs.sche.GetName(), "create-operator-fail").Inc()
+		return nil, nil
+	}
 
-		if bs.rwTy == statistics.Read && bs.cur.region.GetLeader().StoreId == srcStoreID { // move read leader
+	return
+}
+
+func (bs *balanceSolver) createReadOperator(region *core.RegionInfo, srcStoreID, dstStoreID uint64) (op *operator.Operator, typ string, err error) {
+	if region.GetStorePeer(dstStoreID) != nil {
```
Comment (nolouch) [inline server/schedulers/hot_region.go:1068]: The old way use `GetStoreVoter`?

File: server/schedulers/hot_region.go [hunk 4]
```diff
@@ -1012,94 +1018,124 @@ func (bs *balanceSolver) isReadyToBuild() bool {
...
-		if bs.cur.region.GetStoreVoter(dstStoreID) == nil {
-			return nil, nil
-		}
-		desc := "transfer-hot-" + bs.rwTy.String() + "-leader"
+	}
+	return
+}
+
+func (bs *balanceSolver) createWriteOperator(region *core.RegionInfo, srcStoreID, dstStoreID uint64) (op *operator.Operator, typ string, err error) {
+	if region.GetStorePeer(dstStoreID) != nil {
```
Comment (nolouch) [inline server/schedulers/hot_region.go:1105]: Same as.

File: server/schedulers/hot_region.go [hunk 5]
```diff
@@ -1012,94 +1018,124 @@ func (bs *balanceSolver) isReadyToBuild() bool {
...
+	var createOperator func(region *core.RegionInfo, srcStoreID, dstStoreID uint64) (op *operator.Operator, typ string, err error)
+	switch bs.rwTy {
+	case statistics.Read:
+		createOperator = bs.createReadOperator
+	case statistics.Write:
+		createOperator = bs.createWriteOperator
+	}
+
+	currentOp, typ, err := createOperator(bs.cur.region, srcStoreID, dstStoreID)
+	if err == nil {
```
Comment (nolouch) [inline server/schedulers/hot_region.go:1049]: Could we omit this judgment?

**Review by nolouch — COMMENTED** [review]
rest LGTM

File: server/schedulers/hot_region.go [hunk 6]
```diff
@@ -660,22 +666,22 @@ func (bs *balanceSolver) isRegionAvailable(region *core.RegionInfo) bool {
...
 		if srcPeer == nil {
-			log.Debug("region does not have a peer on source store, maybe stat out of date", zap.Uint64("region-id", bs.cur.srcPeerStat.ID()))
+			log.Debug("region does not have a peer on source store, maybe stat out of date", zap.Uint64("region-id", storeID))
 			return nil
 		}
 	case transferLeader:
-		if region.GetLeader().GetStoreId() != bs.cur.srcStore.GetID() {
-			log.Debug("region leader is not on source store, maybe stat out of date", zap.Uint64("region-id", bs.cur.srcPeerStat.ID()))
+		if region.GetLeader().GetStoreId() != storeID {
+			log.Debug("region leader is not on source store, maybe stat out of date", zap.Uint64("region-id", storeID))
```
Comment (bufferflies) [inline server/schedulers/hot_region.go:684]:
```suggestion
			log.Debug("region leader is not on source store, maybe stat out of date", zap.Uint64("leader-store-id", storeID))
```

File: server/schedulers/hot_region.go [hunk 7]
```diff
@@ -563,10 +569,8 @@ func (bs *balanceSolver) checkSrcByDimPriorityAndTolerance(minLoad, expectLoad *
...
 // The returned hotPeer count in controlled by `max-peer-number`.
-func (bs *balanceSolver) filterHotPeers() []*statistics.HotPeerStat {
-	ret := bs.cur.srcStore.HotPeers
-	// Return at most MaxPeerNum peers, to prevent balanceSolver.solve() too slow.
-	maxPeerNum := bs.sche.conf.GetMaxPeerNumber()
+func (bs *balanceSolver) filterHotPeers(storeLoad *statistics.StoreLoadDetail) []*statistics.HotPeerStat {
+	ret := storeLoad.HotPeers
 
 	// filter pending region
 	appendItem := func(items []*statistics.HotPeerStat, item *statistics.HotPeerStat) []*statistics.HotPeerStat {
```
Comment (bufferflies) [inline server/schedulers/hot_region.go:576]: how about refactor to `func(item *statistics.HotPeerStat)[]*statistics.HotPeerStat`?

File: server/schedulers/hot_region.go (hunk 3 above)
Comment (HunDunDM) [reply server/schedulers/hot_region.go:1068]: There are mixed cases in the original code, which are unified into `GetStorePeer`. Considering that `dstStore` is a `Learner`, neither `transfer-leader` nor `move-peer` can be successfully created, so there is no difference.

File: server/schedulers/hot_region.go (hunk 5 above)
Comment (HunDunDM) [reply server/schedulers/hot_region.go:1049]: In subsequent changes, the code in `if err == nil` here may modify `err`.

File: server/schedulers/hot_region.go [hunk 8]
```diff
@@ -1012,94 +1021,124 @@ func (bs *balanceSolver) isReadyToBuild() bool {
...
+
+	currentOp, typ, err := createOperator(bs.cur.region, srcStoreID, dstStoreID)
+	if err == nil {
+		bs.decorateOperator(currentOp, sourceLabel, targetLabel, typ, dim)
+		ops = []*operator.Operator{currentOp}
+		infl = &statistics.Influence{
+			Loads: append(bs.cur.srcPeerStat.Loads[:0:0], bs.cur.srcPeerStat.Loads...),
+			Count: 1,
+		}
+	}
```
Comment (bufferflies) [inline server/schedulers/hot_region.go:1059]:
```suggestion
	}else{
	}
```

**Review by nolouch — APPROVED** [review]
lgtm

File: server/schedulers/hot_region.go (hunk 8 above)
Comment (HunDunDM) [reply server/schedulers/hot_region.go:1059]: Only part of the code is included here, and the `err` in `if err == nil` will be modified in subsequent PRs.

**Pull request merged**